File: 0468-validate-ip-address/0468-validate-ip-address.py
class Solution:
    def validIPAddress(self, queryIP: str) -> str:
        sectors = queryIP.split('.') if '.' in queryIP else queryIP.split(':')
        if (len(sectors) == 4):
            for sec in sectors:
                if not(sec.isdigit() and 0<=int(sec)<=255 and not (sec[0] == '0' and len(sec) >1)):
                    return "Neither"
            return "IPv4"
        elif (len(sectors) == 8):
            for sec in sectors:
                if not(all(c.isdigit() or c.lower() in 'abcdef' for c in sec) and len(sec)<=4 and sec):
                    return "Neither"
            return "IPv6"

        return "Neither"

# there will only be mock ip4 or ip6 therefore the first line can be done without considering other characters
# for ip4 check if the numbers are numbers and within range, and check if there are leading 0 strings
# for ip6 check if the numbers are hex, can be done by checking if each character is either number or within abcdef
# for ip6 also check for section length, to be less than or equal 4, and check for empty sections

File: 0468-validate-ip-address/0468-validate-ip-address.cpp
class Solution {
public:
    int getCount(string source, char target){
        int count = 0;

        for (char c : source) {
            if (c == target) {
                count++;
            }
        }
        return count;

    }
    string validIPAddress(string queryIP) {
        vector<string> sectors;
        string sector;
        stringstream stream(queryIP);
        if (queryIP.empty()){
            return "Neither";
        }

        if (queryIP.find('.') != string::npos && getCount(queryIP, '.') == 3){
            while (getline(stream, sector, '.')){
                sectors.push_back(sector);
            }

            if (sectors.size() == 4){
                for (auto sec : sectors){
                    if (!(all_of( sec.begin(), sec.end(), ::isdigit))){
                        return "Neither";
                    }
                    if (sec.empty() || sec.size() > 4 || (sec.size() > 1 && sec[0] == '0') || stoi(sec) < 0 || stoi(sec) >255){
                        return "Neither";
                    }
                }
                return "IPv4";
            }
        }
        else if (queryIP.back() != ':' && getCount(queryIP, ':') == 7){
            while (getline(stream, sector, ':')){
                sectors.push_back(sector);
            }
            if (sectors.size() == 8){
                string validChars = "0123456789abcdefABCDEF";
                for (auto sec : sectors){
                    if (sec.size() > 4 || sec.empty()){
                        return "Neither";
                    }
                    if (!(all_of( sec.begin(), sec.end(), [validChars](char c) {return validChars.find(c) != string::npos;} ))){
                        return "Neither";
                    }
                }
                return "IPv6";
            }
        }
        return "Neither";
    }
};

// intuition
// 1) a way to split the numbers from the . or :
// 2) a way to check if all characters in a string is acording to a certain condition
// 1 can be done via the combination of find, and getline, find returns the total occurences of a sub string
// and getline will return the substring before a specified character
// 2 can be done via all_of, either by passing a function, or defining your own function by declaring variables inside all_of
// Due to the amount of edge cases in this question, I won't be listing all of them out and how I solved them